Add armv5te-none-eabi and thumbv5te-none-eabi targets (#101329)
rust-lang/rust: src/doc/rustc/src/platform-support/armv5te-none-eabi.md

# `armv5te-none-eabi`

**Tier: 3**

Bare-metal target for any cpu in the ARMv5TE architecture family, supporting
ARM/Thumb code interworking (aka `a32`/`t32`), with ARM code as the default code
generation.

> The `thumbv5te-none-eabi` target is the same as this one, but with THUMB code as the default.

Reviewer: Rust supports switching the “mode” via the (currently unstable) `instruction_set` attribute, so I would argue that describing this in terms of that attribute would make the intent here clearer:

```suggestion
The `thumbv5te-none-eabi` target is the same as this one, but the instruction set defaults to `t32` instead of `a32`.
```

> In particular this supports the main CPU of the Nintendo DS, but there's nothing DS

Reviewer: This sentence is probably superfluous/excessive for the documentation of the very generic baremetal target that `*-none-*` tend to be. There are a wide variety of applications in which v5te has been used (phones, PDAs, etc. and more recently a Kindle which is also a common homebrew/custom firmware target.)

Author: That's fair, the DS specific parts can be removed. I was mostly basing this off of `armv4t-none-eabi`, which makes particular note of the GBA.

according to the specific device you want to target. Pass
`-Clink-arg=-Tyour_script.ld` as a rustc argument to make the linker use
`your_script.ld` during linking.

## Building Rust Programs

Because it is Tier 3, rust does not yet ship pre-compiled artifacts for this target.

Just use the `build-std` nightly cargo feature to build the `core` library. You
can pass this as a command line argument to cargo, or your `.cargo/config.toml`
file might include the following lines:

```toml
[unstable]
build-std = ["core"]
```

Most of `core` should work as expected, with the following notes:
* the target is "soft float", so `f32` and `f64` operations are emulated in
  software.
* integer division is also emulated in software.
* the target is old enough that it doesn't have atomic instructions.

`alloc` is also supported, as long as you provide your own global allocator.

Rust programs are output as ELF files.

For running on DS hardware, you'll need to use an external tool to bundle this ELF file into an NDS binary. The `ndstool` utility included with devkitARM is one such tool that can do this for you:

```shell
ndstool -c [out_nds] -9 [in_elf]

> ```

Reviewer: Similarly this DS-specific note is probably overly specific for the generic target that `*v5te-none-eabi` is.
